File: android/app/src/main/java/io/github/okiyashko1337/felicitydashboard/AjaxActivityClient.java
package io.github.okiyashko1337.felicitydashboard;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.net.SocketTimeoutException;
import java.net.URI;
import java.security.MessageDigest;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.TimeZone;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/** Fast, metadata-only Ajax archive activity query over an ONVIF replay URI. */
final class AjaxActivityClient {
    private final String base,host,user,password;private final int port;private Socket socket;private InputStream input;private OutputStream output;

    AjaxActivityClient(String uri,String user,String password){URI parsed=URI.create(uri);host=parsed.getHost();port=parsed.getPort()>0?parsed.getPort():554;base="rtsp://"+host+(port==554?"":":"+port)+parsed.getRawPath();this.user=user;this.password=password;}

    List<AjaxMetadataDecoder.Activity> fetch(long startMs,long endMs)throws Exception{
        ArrayList<AjaxMetadataDecoder.Activity> found=new ArrayList<>();
        try{
            socket=new Socket(host,port);socket.setSoTimeout(1200);input=new BufferedInputStream(socket.getInputStream(),65536);output=new BufferedOutputStream(socket.getOutputStream());
            Response first=request("DESCRIBE",base,1,"Accept: application/sdp\r\n",null);if(first.auth==null)throw new Exception("Digest challenge missing");
            Response described=request("DESCRIBE",base,2,"Accept: application/sdp\r\n",digest(first.auth,"DESCRIBE",base));if(described.code!=200)throw new Exception("DESCRIBE "+described.code);
            String track=metadataTrack(described.body);Response setup=request("SETUP",track,3,"Transport: RTP/AVP/TCP;unicast;interleaved=0-1\r\nRequire: onvif-replay\r\n",digest(first.auth,"SETUP",track));if(setup.code!=200)throw new Exception("SETUP "+setup.code);
            String session=match(setup.head,"Session:\\s*([^;\\r\\n]+)");if(session==null)throw new Exception("Metadata session missing");
            String headers="Session: "+session+"\r\nRange: clock="+clock(startMs)+"-"+clock(endMs)+"\r\nRate-Control: no\r\nX-Ajax-Metadata-Filter: A\r\nRequire: onvif-replay\r\n";
            Response play=request("PLAY",base,4,headers,digest(first.auth,"PLAY",base));if(play.code!=200)throw new Exception("PLAY "+play.code);
            ByteArrayOutputStream frame=new ByteArrayOutputStream();
            try{
                while(true){int marker=input.read();if(marker<0)break;if(marker!='$')continue;int channel=input.read(),hi=input.read(),lo=input.read();if(lo<0)break;byte[] packet=readFully((hi<<8)|lo);if(channel==1&&isRtcpBye(packet))break;if(channel!=0)continue;int offset=payloadOffset(packet),end=payloadEnd(packet,offset);if(offset<0||offset>=end)continue;frame.write(packet,offset,end-offset);if((packet[1]&0x80)!=0){found.addAll(AjaxMetadataDecoder.decodeActivitiesXml(frame.toString("UTF-8")));frame.reset();}}
            }catch(SocketTimeoutException complete){/* Ajax sends the closed range as a compact batch, then goes idle. */}
        }finally{close();}
        Collections.sort(found,(left,right)->Long.compare(left.timeMs,right.timeMs));return deduplicate(found);
    }

    private static List<AjaxMetadataDecoder.Activity> deduplicate(List<AjaxMetadataDecoder.Activity> source){ArrayList<AjaxMetadataDecoder.Activity> result=new ArrayList<>();for(AjaxMetadataDecoder.Activity candidate:source){AjaxMetadataDecoder.Activity previous=result.isEmpty()?null:result.get(result.size()-1);if(previous!=null&&previous.typeMask==candidate.typeMask&&Math.abs(candidate.timeMs-previous.timeMs)<=250)continue;result.add(candidate);}return result;}
    private Response request(String method,String uri,int cseq,String extra,String auth)throws Exception{StringBuilder value=new StringBuilder(method+" "+uri+" RTSP/1.0\r\nCSeq: "+cseq+"\r\nUser-Agent: Felicity-Activity/1\r\n");if(auth!=null)value.append("Authorization: ").append(auth).append("\r\n");value.append(extra).append("\r\n");output.write(value.toString().getBytes("ISO-8859-1"));output.flush();ByteArrayOutputStream head=new ByteArrayOutputStream();int state=0,next;while((next=input.read())>=0){head.write(next);state=(state==0&&next=='\r')?1:(state==1&&next=='\n')?2:(state==2&&next=='\r')?3:(state==3&&next=='\n')?4:0;if(state==4)break;}String header=head.toString("ISO-8859-1"),length=match(header,"Content-Length:\\s*(\\d+)");byte[] body=length==null?new byte[0]:readFully(Integer.parseInt(length));String status=match(header,"RTSP/1.0\\s+(\\d+)");return new Response(status==null?0:Integer.parseInt(status),header,new String(body,"ISO-8859-1"),match(header,"WWW-Authenticate:\\s*Digest\\s+([^\\r\\n]+)"));}
    private String metadataTrack(String s)throws Exception{boolean metadata=false;for(String raw:s.replace("\r","").split("\n")){String line=raw.trim();if(line.startsWith("m="))metadata=line.startsWith("m=application");else if(metadata&&line.startsWith("a=control:")){String control=line.substring(10).trim();if(control.startsWith("rtsp://"))return control;if(control.startsWith("/"))return "rtsp://"+host+(port==554?"":":"+port)+control;return base+(base.endsWith("/")?"":"/")+control;}}throw new Exception("Metadata control missing from SDP");}
    private static int payloadOffset(byte[] packet){if(packet.length<12||(packet[0]&0xc0)!=0x80)return -1;int offset=12+4*(packet[0]&15);if((packet[0]&0x10)!=0){if(offset+4>packet.length)return -1;offset+=4+((((packet[offset+2]&255)<<8)|(packet[offset+3]&255))*4);}return offset;}
    private static int payloadEnd(byte[] packet,int offset){if(offset<0)return -1;int padding=(packet[0]&0x20)!=0&&packet.length>offset?packet[packet.length-1]&255:0;return padding<=packet.length-offset?packet.length-padding:-1;}
    private static boolean isRtcpBye(byte[] packet){for(int offset=0;offset+4<=packet.length;){int type=packet[offset+1]&255,length=((((packet[offset+2]&255)<<8)|(packet[offset+3]&255))+1)*4;if(type==203)return true;if(length<4||offset+length>packet.length)return false;offset+=length;}return false;}
    private byte[] readFully(int length)throws Exception{byte[] result=new byte[length];int offset=0;while(offset<length){int count=input.read(result,offset,length-offset);if(count<0)throw new Exception("Unexpected RTSP EOF");offset+=count;}return result;}
    private String digest(String challenge,String method,String address)throws Exception{String realm=param(challenge,"realm"),nonce=param(challenge,"nonce"),ha1=md5(user+":"+realm+":"+password),ha2=md5(method+":"+address);return "Digest username=\""+user+"\", realm=\""+realm+"\", nonce=\""+nonce+"\", uri=\""+address+"\", response=\""+md5(ha1+":"+nonce+":"+ha2)+"\"";}
    private void close(){try{if(socket!=null)socket.close();}catch(Exception ignored){}socket=null;}
    private static String clock(long time){SimpleDateFormat f=new SimpleDateFormat("yyyyMMdd'T'HHmmss'Z'",Locale.US);f.setTimeZone(TimeZone.getTimeZone("UTC"));return f.format(new Date(time));}
    private static String param(String s,String name){Matcher m=Pattern.compile(name+"=\"([^\"]+)\"",Pattern.CASE_INSENSITIVE).matcher(s);return m.find()?m.group(1):"";}
    private static String match(String s,String re){Matcher m=Pattern.compile(re,Pattern.CASE_INSENSITIVE).matcher(s);return m.find()?m.group(1):null;}
    private static String md5(String s)throws Exception{byte[] bytes=MessageDigest.getInstance("MD5").digest(s.getBytes("ISO-8859-1"));StringBuilder out=new StringBuilder();for(byte b:bytes)out.append(String.format(Locale.US,"%02x",b&255));return out.toString();}
    private static final class Response{final int code;final String head,body,auth;Response(int code,String head,String body,String auth){this.code=code;this.head=head;this.body=body;this.auth=auth;}}
}

File: ios/FelicityDashboard/AjaxActivityClient.swift
import Foundation

actor AjaxActivityClient {
    private let replayService: ProfileGArchiveService

    init(replayService: ProfileGArchiveService = .shared) { self.replayService = replayService }

    func fetch(
        camera: CameraDescriptor,
        quality: StreamQuality,
        configuration: RecorderConfiguration,
        start: Date,
        end: Date
    ) async throws -> [ArchiveInterval] {
        let replay = try await replayService.replay(camera: camera, quality: quality, configuration: configuration)
        guard let url = URL(string: replay.uri), let host = url.host else { throw RTSPError.invalidURI }
        let client = ArchiveRTSPClient()
        let accumulator = MetadataAccumulator()
        try await client.connect(host: host, port: UInt16(url.port ?? 554)) { channel, data in
            Task { await accumulator.accept(channel: channel, packet: data) }
        }
        defer { Task { await client.close() } }

        let first = try await client.request("DESCRIBE", target: replay.uri, headers: ["Accept: application/sdp"], userAgent: "Felicity-Activity/1")
        let challenge = first.headers["www-authenticate"] ?? ""
        let described: ArchiveRTSPResponse
        if first.code == 401 {
            guard !challenge.isEmpty else { throw RTSPError.authentication }
            described = try await client.request(
                "DESCRIBE", target: replay.uri, headers: ["Accept: application/sdp"],
                challenge: challenge, username: configuration.username, password: configuration.password,
                userAgent: "Felicity-Activity/1"
            )
        } else { described = first }
        guard described.code == 200 else { throw ArchiveError.response("Metadata DESCRIBE \(described.code)") }
        let track = try Self.metadataTrack(described.body, contentBase: described.headers["content-base"] ?? replay.uri)
        let setup = try await client.request(
            "SETUP", target: track,
            headers: ["Transport: RTP/AVP/TCP;unicast;interleaved=0-1", "Require: onvif-replay"],
            challenge: challenge, username: configuration.username, password: configuration.password,
            userAgent: "Felicity-Activity/1"
        )
        guard setup.code == 200 else { throw ArchiveError.response("Metadata SETUP \(setup.code)") }
        let session = setup.headers["session"]?.components(separatedBy: ";").first?.trimmingCharacters(in: .whitespaces) ?? ""
        guard !session.isEmpty else { throw ArchiveError.response("Metadata session missing") }
        let play = try await client.request(
            "PLAY", target: replay.uri,
            headers: [
                "Session: \(session)",
                "Range: clock=\(ReplayClock.clock(start))-\(ReplayClock.clock(end))",
                "Rate-Control: no",
                "X-Ajax-Metadata-Filter: A",
                "Require: onvif-replay",
            ],
            challenge: challenge, username: configuration.username, password: configuration.password,
            userAgent: "Felicity-Activity/1"
        )
        guard play.code == 200 else { throw ArchiveError.response("Metadata PLAY \(play.code)") }

        for _ in 0..<40 {
            try await Task.sleep(for: .milliseconds(50))
            if await accumulator.complete { break }
            if await accumulator.receivedAny, await accumulator.idleFor >= 0.25 { break }
        }
        let boundaries = await accumulator.boundaries
        return ArchiveTimelineRules.intervals(from: boundaries)
    }

    private static func metadataTrack(_ sdp: String, contentBase: String) throws -> String {
        var inMetadata = false
        for raw in sdp.replacingOccurrences(of: "\r", with: "").split(separator: "\n", omittingEmptySubsequences: false) {
            let line = raw.trimmingCharacters(in: .whitespaces)
            if line.hasPrefix("m=") { inMetadata = line.hasPrefix("m=application") }
            else if inMetadata, line.hasPrefix("a=control:") {
                let control = String(line.dropFirst("a=control:".count)).trimmingCharacters(in: .whitespaces)
                if control.lowercased().hasPrefix("rtsp://") { return control }
                if control.hasPrefix("/"), let base = URL(string: contentBase), let scheme = base.scheme, let host = base.host {
                    return "\(scheme)://\(host)\(base.port.map { ":\($0)" } ?? "")\(control)"
                }
                return contentBase.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + "/" + control
            }
        }
        throw ArchiveError.response("Ajax activity metadata track is missing")
    }
}

private actor MetadataAccumulator {
    private var xml = Data()
    private(set) var boundaries: [ArchiveActivityBoundary] = []
    private(set) var complete = false
    private(set) var receivedAny = false
    private var lastReceived = Date.distantPast

    var idleFor: TimeInterval { Date().timeIntervalSince(lastReceived) }

    func accept(channel: Int, packet: Data) {
        if channel == 1 {
            if Self.containsRTCPBye(packet) { complete = true }
            return
        }
        guard channel == 0, let payload = Self.rtpPayload(packet) else { return }
        receivedAny = true
        lastReceived = .now
        xml.append(payload)
        guard packet.count > 1, packet[packet.startIndex + 1] & 0x80 != 0 else { return }
        if let text = String(data: xml, encoding: .utf8) { boundaries.append(contentsOf: AjaxActivityDecoder.decodeXML(text)) }
        xml.removeAll(keepingCapacity: true)
    }

    private static func rtpPayload(_ packet: Data) -> Data? {
        let bytes = [UInt8](packet)
        guard bytes.count >= 12, bytes[0] & 0xc0 == 0x80 else { return nil }
        var offset = 12 + Int(bytes[0] & 0x0f) * 4
        guard offset <= bytes.count else { return nil }
        if bytes[0] & 0x10 != 0 {
            guard offset + 4 <= bytes.count else { return nil }
            let words = Int(bytes[offset + 2]) << 8 | Int(bytes[offset + 3])
            offset += 4 + words * 4
        }
        let padding = bytes[0] & 0x20 != 0 ? Int(bytes.last ?? 0) : 0
        let end = bytes.count - padding
        guard offset < end else { return nil }
        return Data(bytes[offset..<end])
    }

    private static func containsRTCPBye(_ packet: Data) -> Bool {
        let bytes = [UInt8](packet)
        var offset = 0
        while offset + 4 <= bytes.count {
            let type = bytes[offset + 1]
            let length = ((Int(bytes[offset + 2]) << 8 | Int(bytes[offset + 3])) + 1) * 4
            if type == 203 { return true }
            guard length >= 4, offset + length <= bytes.count else { return false }
            offset += length
        }
        return false
    }
}
